### PR TITLE
[SPARK-42283][CONNECT][SCALA] Simple Scalar Scala UDFs

### PR DESCRIPTION
**Pull request by vicennial**
### What changes were proposed in this pull request?

This PR adds support for "simple" scalar Scala UDFs for the Spark Connect Scala/JVM Client. “Simple” here refers to UDFs that utilize no client-specific class files (e.g REPL-generated) and JARs. Essentially, a “simple” UDF may only reference in-built libraries and classes defined within the scope of the UDF.

A user would then be able to do the following (example):
```
def myFunc(x: Int): Int = x + 5

val myUdf = udf(myFunc _)
df = df.select(myUdf(Column("id")))
```
#### Implementation Details:

A shared JVM object `UdfPacket` is introduced in the common package to encapsulate the Scala UDF and its encoders (via Agnostic Encoders) such that it could be serialized/deserialized on the client and server respectively. 

Further, a new protobuf message `ScalarScalaUDF` is introduced to transmit Scala/JVM specific information to the server (such as the above serialized JVM object).

### Why are the changes needed?

UDFs are crucial for the completeness of the Spark Connect Scala/JVM client. We introduce this component incrementally.

### Does this PR introduce _any_ user-facing change?
Yes, users are now able to run "simple" scalar Scala UDFs through the Scala/JVM client.

### How was this patch tested?
Unit test + Integration test

### CODE REVIEW

File: connector/connect/common/pom.xml [hunk 1]
```diff
@@ -38,6 +38,18 @@
         <tomcat.annotations.api.version>6.0.53</tomcat.annotations.api.version>
     </properties>
     <dependencies>
+        <dependency>
```
Comment (hvanhovell) [inline connector/connect/common/pom.xml:41]: Why do we need to add this?

Comment (vicennial) [reply connector/connect/common/pom.xml:41]: For the shared `UdfPacket` (defined in `common`) containing `AgnosticEncoder` related params

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 2]
```diff
@@ -319,3 +320,14 @@ message PythonUDF {
   bytes command = 3;
 }
 
+message ScalarScalaUDF {
+  // (Required) Serialized JVM object containing UDF definition, input encoders and output encoder
+  bytes payload = 1;
+  // (Optional) Input type(s) of the UDF
+  repeated DataType inputTypes = 2;
+  // (Required) Output type of the UDF
+  DataType outputType = 3;
```
Comment (hvanhovell) [inline connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:329]: @xinrong-meng @vicennial can we homogenize this and put the return type in the `CommonInlineUserDefinedFunction` message?

File: connector/connect/common/src/main/protobuf/spark/connect/expressions.proto [hunk 3]
```diff
@@ -319,3 +320,14 @@ message PythonUDF {
...
+  // (Required) True if the UDF can return null value
+  bool nullable = 4;
```
Comment (hvanhovell) [inline connector/connect/common/src/main/protobuf/spark/connect/expressions.proto:331]: Same question.

**Review by hvanhovell — APPROVED** [review]
LGTM

File: connector/connect/common/pom.xml (hunk 1 above)
Comment (hvanhovell) [reply connector/connect/common/pom.xml:41]: Sorry I am blind. Yeah that makes sense.

**Comment by hvanhovell** [comment]
Merging this to master/3.4

**Comment by LuciferYang** [comment]
I write a demo code out of Spark project to test SPARK-42283, the demo code in 

https://github.com/LuciferYang/SparkConnectDemo

Build a spark client and Use the following commands to start a connect server:

```
bin/spark-shell --jars spark-connect_2.12-3.5.0-SNAPSHOT.jar --driver-class-path spark-connect_2.12-3.5.0-SNAPSHOT.jar --conf spark.plugins=org.apache.spark.sql.connect.SparkConnectPlugin --conf spark.connect.grpc.binding.port=15102

```

Then when I run `ConnectDemo` or `ConnectDemoSuite`, I will found the follwoing error in the server side:

```
scala> 23/02/02 18:02:46 ERROR SparkConnectService: Error during: execute
java.lang.ClassNotFoundException: personal.code.ConnectDemo$
	at scala.reflect.internal.util.AbstractFileClassLoader.findClass(AbstractFileClassLoader.scala:72)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.apache.spark.util.Utils$$anon$1.resolveClass(Utils.scala:142)
	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1988)
	at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1852)
	at java.io.ObjectInputStream.readClass(ObjectInputStream.java:1815)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1640)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2431)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2355)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2213)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1669)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2431)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2355)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2213)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1669)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:503)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:461)
	at org.apache.spark.util.Utils$.deserialize(Utils.scala:146)
	at org.apache.spark.sql.connect.planner.SparkConnectPlanner.transformScalarScalaUDF(SparkConnectPlanner.scala:855)
	at org.apache.spark.sql.connect.planner.SparkConnectPlanner.transformCommonInlineUserDefinedFunction(SparkConnectPlanner.scala:837)
	at org.apache.spark.sql.connect.planner.SparkConnectPlanner.transformExpression(SparkConnectPlanner.scala:748)
	at org.apache.spark.sql.connect.planner.SparkConnectPlanner.$anonfun$transformProject$1(SparkConnectPlanner.scala:698)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at scala.collection.TraversableLike.map(TraversableLike.scala:286)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:279)
	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
	at org.apache.spark.sql.connect.planner.SparkConnectPlanner.transformProject(SparkConnectPlanner.scala:698)
	at org.apache.spark.sql.connect.planner.SparkConnectPlanner.transformRelation(SparkConnectPlanner.scala:72)
	at org.apache.spark.sql.connect.service.SparkConnectStreamHandler.handlePlan(SparkConnectStreamHandler.scala:58)
	at org.apache.spark.sql.connect.service.SparkConnectStreamHandler.handle(SparkConnectStreamHandler.scala:49)
	at org.apache.spark.sql.connect.service.SparkConnectService.executePlan(SparkConnectService.scala:135)
	at org.apache.spark.connect.proto.SparkConnectServiceGrpc$MethodHandlers.invoke(SparkConnectServiceGrpc.java:306)
	at org.sparkproject.connect.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:182)
	at org.sparkproject.connect.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:352)
	at org.sparkproject.connect.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:866)
	at org.sparkproject.connect.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at org.sparkproject.connect.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
```

Are there problems with my test way? @vicennial @hvanhovell 

The test code is similar as `simple udf test`:

```scala
    val spark = SparkSession.builder()
      .client(SparkConnectClient.builder().port(15102).build()).build()
    try {
      def dummyUdf(x: Int): Int = x + 5
      val myUdf = udf(dummyUdf _)
      val df = spark.range(5).select(myUdf(Column("id")))

      val result = df.collectResult()
      assert(result.length == 5)
      result.toArray.zipWithIndex.foreach { case (v, idx) =>
        assert(v.getInt(0) == idx + 5)
      }
    } finally {
      spark.close()
    }
```

**Comment by LuciferYang** [comment]
Then I try to do this demo in spark shell:

1. build a spark-client
2. add `spark-connect-client-jvm_2.12-3.5.0-SNAPSHOT.jar` and `perfmark-api-0.25.0.jar` into  `jars`
3. remove `spark-sql_2.12-3.5.0-SNAPSHOT.jar ` from `jars` due to package+classname conflict between `spark-connect-client-jvm_2.12-3.5.0-SNAPSHOT.jar`  and `spark-sql_2.12-3.5.0-SNAPSHOT.jar `
4. start client shell with `bin/spark-shell`, it is usable for spark-connect-client-jvm_2.12-3.5.0-SNAPSHOT.jar` although there are the following errors:

```
<console>:18: error: value sparkContext is not a member of org.apache.spark.sql.SparkSession
             val _sc = spark.sparkContext
                             ^
<console>:14: error: object implicits is not a member of package spark
       import spark.implicits._
                    ^
<console>:14: error: object sql is not a member of package spark
       import spark.sql
              ^
```
5. run the following code in shell:
```

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.connect.client.SparkConnectClient
import org.apache.spark.sql.Column
import org.apache.spark.sql.functions.udf
val ss = SparkSession.builder().client(SparkConnectClient.builder().port(15102).build()).build()
def dummyUdf(x: Int): Int = x + 5
val myUdf = udf(dummyUdf _)
val df = ss.range(5).select(myUdf(Column("id")))
val result = df.collectResult()
println(result.length)
result.toArray.zipWithIndex.foreach { case (v, idx) =>
  println(v.getInt(0))
}
```

then we can see following error in server side:

```
23/02/02 18:44:00 ERROR SparkConnectService: Error during: execute
java.lang.ClassNotFoundException: $line14.$read$$iw$$iw$$iw$$iw$$iw$$iw
	at scala.reflect.internal.util.AbstractFileClassLoader.findClass(AbstractFileClassLoader.scala:72)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.apache.spark.util.Utils$$anon$1.resolveClass(Utils.scala:142)
	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1988)
	at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1852)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2186)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1669)
	at java.io.ObjectInputStream.readArray(ObjectInputStream.java:2119)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1657)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2431)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2355)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2213)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1669)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2431)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2355)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2213)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1669)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:503)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:461)
	at org.apache.spark.util.Utils$.deserialize(Utils.scala:146)
	at org.apache.spark.sql.connect.planner.SparkConnectPlanner.transformScalarScalaUDF(SparkConnectPlanner.scala:855)
	at org.apache.spark.sql.connect.planner.SparkConnectPlanner.transformCommonInlineUserDefinedFunction(SparkConnectPlanner.scala:837)
	at org.apache.spark.sql.connect.planner.SparkConnectPlanner.transformExpression(SparkConnectPlanner.scala:748)
	at org.apache.spark.sql.connect.planner.SparkConnectPlanner.$anonfun$transformProject$1(SparkConnectPlanner.scala:698)
```

**Comment by hvanhovell** [comment]
@LuciferYang thanks for trying this out! The UDF code doesn't support REPL class sync yet. That will be out in a couple of days and then this should work :).

**Comment by hvanhovell** [comment]
As for the first error, the classes also have to be on the 'server' side.

**Pull request closed without merging**